qiskit/circuit/instructionset.py

```diff
@@ -142,24 +141,26 @@ def __init__(self, circuit_cregs=None, *, resource_requester: Optional[Callable]
 
     def __len__(self):
         """Return number of instructions in set"""
-        return len(self.instructions)
+        return len(self._instructions)
 
     def __getitem__(self, i):
         """Return instruction at index"""
-        return self.instructions[i]
+        return self._instructions[i]
 
-    def add(self, gate, qargs, cargs):
+    def add(self, instruction, qargs=None, cargs=None):
         """Add an instruction and its context (where it is attached)."""
-        if not isinstance(gate, Instruction):
-            raise CircuitError("attempt to add non-Instruction" + " to InstructionSet")
-        self.instructions.append(gate)
-        self.qargs.append(qargs)
-        self.cargs.append(cargs)
+        if not isinstance(instruction, CircuitInstruction):
+            if not isinstance(instruction, Instruction):
+                raise CircuitError("attempt to add non-Instruction to InstructionSet")
+            if qargs is None or cargs is None:
```

Comment:
We could do.  I was kind of avoiding anything that involved additional checks when using the preferred interface.  Having the preferred form override any specifications of the old form is how we handle the `BackendV1` or `BackendV2`/`Target` split in the transpiler too, but I'm ok if you'd rather shout louder in this case.  My main thought was just about not adding extra checks in the preferred path for performance reasons, because this gets called on all appends to `QuantumCircuit`.